video: Add ewa_lanczos upscaler (aka jinc)

haasn opened the pull request:
This is the polar (elliptic weighted average) version of lanczos.
This introduces a general new form of polar/2D filters.

This commit probably still needs a bit of work before it can be merged, but I've created a pull request to keep track of feedback and make it visible.

haasn commented:
Seems like this version is actually slower than the previous hacked-together version. The culprit seems to be the usage of sampler3D instead of sampler2D.

Based on this, I tried going even smaller and using a sampler1D that just caches the jinc function (with the weights matrix computed in realtime on the GPU). Turned out to be even faster.

Clearly, this still needs some work.

bjin commented on video/out/filter_kernels.c:
```diff
@@ -303,6 +383,10 @@ const struct filter_kernel mp_filter_kernels[] = {
     {"sinc3",          3,   sinc},
     {"sinc4",          4,   sinc},
     {"sinc",           -1,  sinc},
+    {"ewa_lanczos16",  2,   ewa_lanczos,    .polar = true},
```
This is inconsistent with other existing filter names. How about jinc2/3/4?

haasn replied:
The general idea is that the number is relative to performance, since ewa_lanczos64 uses 64 point samples (whereas lanczos2 only uses 2 per direction), but I guess it can really be argued either way.

One point of interest is that this reasoning is inconsistent with spline16/32/64, since that's also based on a separable kernel and hence indistinguishable from lanczos2/3/4 in terms of performance.

I think you're right, though, it should be named based on the number of taps.

wm4 replied:
We could easily fix the other entries in a backward compatible way by adding duplicate entries.

haasn commented:
I've pushed a rework of this commit that uses a much faster 1D LUT approach. With this version, I can actually watch 60 Hz 4K videos in realtime even with ewa_lanczos4, which is a quite impressive improvement over the previous, extremely slow version. Speedup is almost 100% on my machine.

wm4 commented:
Does using a 1D texture vs. 2D texture make a difference?

haasn commented:
@wm4 Depends on what you mean:

1. Using a 2D texture for the LUT (with x and y each mapped to a coordinate) is slower than using a 1D texture (with sqrt(xx+yy) mapped to a coordinate).

The performance numbers I get for this on my machine is something like: 12 ms (1D), 15 ms (2D) and 19ms (3D).

2. Using a 2D texture of size Nx1 to simulate a 1D texture seems to be even slower, I get about 17ms avg GPU time. (Even slower if I use 1xN, for some reason)

wm4 commented:
I've recently replaced the 1D LUTs for the separable scalers with a 1xN LUT (1 pixel width), and found no measurable speed difference. This was on a GeForce 8500 GT with nvidia drivers.

haasn commented:
Which scalers does that refer to? None of the separable scalers in the current code use 1D LUTs, at least as far as I can see.

(I've rebased and squashed the commit, for easier access to the new code)

wm4 commented:
>Which scalers does that refer to? None of the separable scalers in the current code use 1D LUTs, at least as far as I can see.

The convolution filters. E.g. ``weights2()`` was implemented using a 1D texture. This was changed in 4a95be014b8875567e8e800a6075d876ba824081. Maybe you could test whether the commit reduces performance for you?

haasn commented:
I ran each test a few times to eliminate margin of error as much as possible. The numbers I'm reporting here are the ones from the first program listed by qapitrace, which, if I'm not mistaken, corresponds to the second pass only for the separable filters (it has 8 avg million pixel drawn, and the highest time taken).

4cae83d, lanczos3: 1.795ms, 1.799ms, 1.792ms, 1.806ms, 1.792ms
4a95be0, lanczos3: 1.799ms, 1.806ms, 1.799ms, 1.799ms, 1.799ms

4cae83d, lanczos2: 1.138ms, 1.313ms, 1.120ms, 1.482ms, 1.120ms
4a95be0, lanczos2: 1.132ms, 1.305ms, 1.156ms, 1.163ms, 1.139ms

I'm not sure what causes the high amount of fluctuations in particular in the lanczos2 case, but it doesn't seem like it significantly impacts performance here. To eliminate any similar margin of error from my other test, here is the exact patch used and the numbers I get out of them.

unpatched, ewa_lanczos4: 11.988ms, 11.949ms, 11.961ms, 11.918ms
unpatched, ewa_lanczos2: 3.793ms, 3.794ms, 3.819ms, 3.839ms, 3.808ms

This is the first version of the patch, for a 1xN texture: https://bpaste.net/show/a80e0fa8e8e5

patched 1xN, ewa_lanczos4: 20.485ms, 20.463ms, 20.502ms, 20.462ms
patched 1xN, ewa_lanczos2: 5.607ms, 5.608ms, 5.619ms, 5.611ms, 5.607ms

This is the second version of the patch, for an Nx1 texture: https://bpaste.net/show/2b0905cfa371

patched Nx1, ewa_lanczos4: 17.303ms, 17.308ms, 17.304ms, 17.310ms
patched Nx1, ewa_lanczos2: 5.181ms, 5.154ms, 5.153ms, 5.158ms, 5.154ms

There is a clearly noticeable speed loss here. Maybe it's related to the number of texture channels? In each case, I am still using monochrome textures (in GL_RED format).

wm4 commented:
Huh, that's really weird. I can't make any sense out of this. Just for the record, what's your gpu/driver?

We could probably just leave it as is, but still makes me feel uneasy, and it makes the code more complex.

haasn commented:
My GPU is AMD Radeon HD 7950 (900 MHz) and the exact version information is:

```
Package: x11-drivers/ati-drivers-14.9-r1
Catalyst version: 12.3
Driver packaging version: 8.951-120308a-135854C-ATI
2D Driver version: 14.30.4
Catalyst control center version: 2.21
RandR version: 1.3
```

(Not really sure how to interpret those version numbers)

haasn commented:
Merged.
